pysph/tools/ipython.py

```diff
@@ -0,0 +1,590 @@
+        'inlet': <pysph.base.particle_array.ParticleArray at 0x7f3f7d144b98>,
+        'outlet': <pysph.base.particle_array.ParticleArray at 0x7f3f7d144c30>},
+        'solver_data': {'count': 240, 'dt': 0.01, 't': 2.399999999999993}}
+
+
+        '''
+
+        if self.cache is not None:
+            if frame in self.cache:
+                temp_data = self.cache[frame]
+            else:
+                self.cache[frame] = temp_data = load(self.paths_list[frame])
+        else:
+            temp_data = load(self.paths_list[frame])
+
+        return temp_data
+
+    def show_log(self):
+        '''
+        Prints the content of log file.
+        '''
+
+        print("Printing log : \n\n")
+        path = self.path + "*.log"
+        with open(glob.glob(path)[0], 'r') as logfile:
+            for lines in logfile:
+                print(lines)
+
+    def show_results(self):
+        '''
+        Show if there are any png, jpeg, jpg, or bmp images.
+        '''
+
+        imgs = tuple()
+        for extension in ['png', 'jpg', 'jpeg', 'bmp']:
+            temppath = self.path + "*." + extension
+            for paths in glob.glob(temppath):
+                imgs += (Image(paths),)
+        if len(imgs) != 0:
+            display(*imgs)
+        else:
+            print("No results to show.")
+
+    def show_info(self):
+        '''
+        Print contents of the .info file present in the output directory,
+        keys present in results.npz,  number of files and
+        information about paricle arrays.
+        '''
+
+        # General Info #
+
+        path = self.path + "*.info"
+        with open(glob.glob(path)[0], 'r') as infofile:
+            data = json.load(infofile)
+
+            print('Printing info : \n')
+            for key in data.keys():
+                if key == 'cpu_time':
+                    print(key + " : " + str(data[key]) + " seconds")
+                else:
+                    print(key + " : " + str(data[key]))
+
+            print('Number of files : {}'.format(len(self.paths_list)))
+
+        # Particle Info #
+
+        temp_data = load(self.paths_list[0])['arrays']
+
+        for key in temp_data:
+            print("  {} :".format(key))
+            print("    Number of particles : {}".format(
+                temp_data[key].get_number_of_particles())
+            )
+            print("    Output Property Arrays : {}".format(
+                temp_data[key].output_property_arrays)
+            )
+
+        # keys in results.npz
+
+        from numpy import load as npl
+
+        path = self.path + "*results*"
+        files = glob.glob(path)
+        if len(files) != 0:
+            data = npl(files[0])
+            print("\nKeys in results.npz :")
+            print(data.keys())
+
+    def show_all(self):
+        self.show_info()
+        self.show_results()
+        self.show_log()
+
+
+class Viewer2DWidgets(object):
+
+    def __init__(self, file, file_count, handlers):
```

Comment:
Still too complex.  You don't need handlers at all, you could as well have passed the Viewer2D object and called observe on `viewer._frame_handler` etc.